grype/pkg/purl_provider.go

```diff
@@ -65,6 +66,18 @@ func decodePurlFile(reader io.Reader) ([]Package, error) {
 					cpes = append(cpes, c)
 				}
 			}
+
+			if qualifier.Key == "epoch" {
+				epoch = qualifier.Value
+			}
+		}
+
+		if err := scanner.Err(); err != nil {
+			return nil, err
+		}
+
+		if !strings.Contains(purl.Version, fmt.Sprintf("%s:", epoch)) {
```

Comment:
Should this possibly check if `epoch != "0"` as well? Is 0 used for an `epoch`, if so maybe change default to `-1`. Then do a similar thing of if `epoch != "-1"` and its not already part of the version it will be added?

Reply:
This change could default `epoch = ""` perhaps so it would only be non-empty string if it was set in the qualifiers. Most importantly, this should _not_ affect versions for ecosystems that don't use the epoch value. As it stands, this will affect _all_ versions, for _all_ ecosystems, I think. It should probably test if `purl.Type == packageurl.TypeRPM` or similar, what do you think?

Reply:
@kzantow Something more like:

```go
if epoch != "" && purl.Type == packageurl.TypeRPM && !strings.Contains(purl.Version, fmt.Sprintf("%s:", epoch)) {
    purl.Version = fmt.Sprintf("%s:%s", epoch, purl.Version)
}
```
?

Reply:
Should an empty epoch be the same as `0`? (I don't know the answer here), would this make more sense?
```go
if purl.Type == packageurl.TypeRPM {
  if epoch == "" {
    epoch = "0" // or this just defaults to "0" as you have it
  }
  if !strings.HasPrefix(purl.Version, fmt.Sprintf("%s:", epoch)) {
    purl.Version = fmt.Sprintf("%s:%s", epoch, purl.Version)
  }
}
```

Reply:
Yeah that makes sense should I just keep `epoch` defaulting to `"0"` so it would be this?

```go
if purl.Type == packageurl.TypeRPM {
  if !strings.HasPrefix(purl.Version, fmt.Sprintf("%s:", epoch)) {
    purl.Version = fmt.Sprintf("%s:%s", epoch, purl.Version)
  }
}
```

Reply:
Combine the if statements, but yeah:

```go
if purl.Type == packageurl.TypeRPM && !strings.HasPrefix(purl.Version, fmt.Sprintf("%s:", epoch)) {
    purl.Version = fmt.Sprintf("%s:%s", epoch, purl.Version)
}
```